### ISSUE TITLE
data source depending on another

### ISSUE THREAD

**Issue by samar-elsayed** [hashicorp/terraform#33562]
### Terraform Version

```shell
Terraform v1.5.2
```

### Terraform Configuration Files

```
resource "time_sleep" "wait_15_seconds" {
  create_duration = "50s"
  depends_on      = [aws_lb.lb]
}

data "aws_lbs" "this" {
  tags = {
    type   = "application"
  }
  depends_on = [
    time_sleep.wait_15_seconds
  ]
}

data "aws_lb" "this" {
  for_each = toset(data.aws_lbs.this.arns)
  arn   = each.key

  depends_on = [
    data.aws_lbs.this
  ]

}
```

### Debug Output

[debug.txt](https://github.com/hashicorp/terraform/files/12129590/debug.txt)

### Expected Behavior

aws_lb should be available so I can create the locals

### Actual Behavior

 Error: Invalid for_each argument
│ 
│   on modules/loadbalancer/main.tf line 202, in data "aws_lb" "this":
│  202:   for_each = toset(data.aws_lbs.this.arns)
│     ├────────────────
│     │ data.aws_lbs.this.arns is a set of string, known only after apply

### Steps to Reproduce

terraform init
terraform apply

**Comment by jbardin** [comment]
Hi @samar-elsayed,

Thanks for filing the issue. Adding a data source to `depends_on` will ensure the order that the data sources are evaluated, however because a data source has no external side effects, adding a data source to `depends_on` won't cause reading of the dependent data source to be deferred.

The actual condition you are waiting for in `data.aws_lb.this` is for the changes in the managed resource to be applied, therefor the address of the managed resource is what should be set in the `depends_on`.

Since this is working as designed, and does not represent a bug in Terraform I'm going to close out the issue.  If you have more questions, it's better to use [the community forum](https://discuss.hashicorp.com/c/terraform-core) where there are more people ready to help.

Thanks!

**Comment by samar-elsayed** [comment]
Hi @jbardin 

Do you mean to have it like that
```

data "aws_lb" "this" {
  for_each = toset(data.aws_lbs.this.arns)
  arn   = each.key
  depends_on = [
    time_sleep.wait_15_seconds
  ]

}
```
Because I tried that too but didn't work

**Comment by jbardin** [comment]
That declares that `data.aws_lb.this` must wait for all changes to `time_sleep.wait_15_seconds` be applied before reading the data source. If that still does not work, then perhaps there is a problem dealing with eventual consistency from the remote system, which is implied by the fact that you need to have `time_sleep` in the first place. Using `time_sleep` like this is usually only needed when the provider responds that the resource has been created before it is in fact ready for use. 

There could also be other complications if you are dealing with updates, because you are using both a data source and a managed resource to represent the same logical resource in the configuration, which is never recommended. A more complete example may help, and someone more familiar with AWS details might be able to assist in the [community forums](https://discuss.hashicorp.com/c/terraform-providers/tf-aws/33)
